Sources/Build/BuildPlan.swift

```diff
@@ -1762,18 +1781,17 @@ public class BuildPlan {
     }
 
     static func validateDeploymentVersionOfProductDependency(
-        _ product: ResolvedProduct,
+        product: ResolvedProduct,
+        productPackage: ResolvedPackage,
         forTarget target: ResolvedTarget,
         observabilityScope: ObservabilityScope
     ) throws {
-        // Get the first target as supported platforms are on the top-level.
-        // This will need to become a bit complicated once we have target-level platform support.
-        let productTarget = product.underlyingProduct.targets[0]
-
-        guard let productPlatform = productTarget.getSupportedPlatform(for: .macOS) else {
-            throw StringError("Expected supported platform macOS in product target \(productTarget)")
+        // Supported platforms are defined at the package level.
+        // This will need to become a bit complicated once we have target-level or product-level platform support.
+        guard let productPlatform = productPackage.platforms.getDerived(for: .macOS) else {
```

Comment:
I think this should continue to check one of the targets of the actual product instead of just the package itself.

Reply:
Hmm... I would agree with that if each target could define its platform version (in which case we'd also need to combine them to arrive at the product's minimum platform version).  But given that the platform is currently defined on the package, it seems clearer to reflect that in the code too.  Once we change it in the manifest, it seems to me that we would then make the internal changes too.

Reply:
I would agree with this if we were also getting rid of the by-target attribute, but we are not, so we are just making things less clear here. Also test targets for one can have a different platform because they'll be made to match whatever the underlying XCTest framework being linked supports. This creates a weird ambiguity of whether that is the correct platform to use or the package one.

Reply:
there previous code has this: `let productTarget = product.underlyingProduct.targets[0]` which I think is less accurate than what is proposed here

Reply:
I agree that this is clearer.  The point about test targets is interesting, though — even so, I still think it's clearer in that case to have `platforms` and `testPlatforms` on the package rather than do the `target[0]` thing.

Reply:
https://github.com/apple/swift-package-manager/pull/4191#discussion_r821025130